File: modules/widgets/lib/rin/widgets/containers/WCFitter.cpp
#include "rin/widgets/containers/WCFitter.hpp"

#include "rin/widgets/ContainerWidgetSlot.hpp"

FitMode WCFitter::GetMode() const
{
    return _mode;
}

void WCFitter::SetMode(const FitMode mode)
{
    const auto oldMode = _mode;
    _mode = mode;
    if (mode != oldMode && GetSurface())
    {
        ArrangeSlots(GetContentSize());
    }
}

Vec2<float> WCFitter::ComputeContentSize()
{
    if (const auto slot = GetSlot(0))
    {
        return slot->GetWidget()->GetDesiredSize();
    }

    return Vec2{0.0f};
}

size_t WCFitter::GetMaxSlots() const
{
    return 1;
}

void WCFitter::ArrangeSlots(const Vec2<float>& drawSize)
{
    SizeContent(GetContentSize());
}

Vec2<float> WCFitter::ComputeContainSize(const Vec2<float>& drawSize, const Vec2<float>& desiredSize)
{
    const auto aspect = desiredSize.y / desiredSize.x;
    const Vec2 scaledWidgetSize{drawSize.x, drawSize.x * aspect};

    if (drawSize.NearlyEquals(scaledWidgetSize, 0.001)) return scaledWidgetSize;

    return scaledWidgetSize.y < drawSize.y ? scaledWidgetSize : Vec2{drawSize.y / aspect, drawSize.y};
}

Vec2<float> WCFitter::ComputeCoverSize(const Vec2<float>& drawSize, const Vec2<float>& desiredSize)
{
    const auto aspect = desiredSize.y / desiredSize.x;
    const Vec2 scaledWidgetSize{drawSize.x, drawSize.x * aspect};

    if (drawSize.NearlyEquals(scaledWidgetSize, 0.001)) return scaledWidgetSize;

    return scaledWidgetSize.y < drawSize.y ? Vec2{drawSize.y / aspect, drawSize.y} : scaledWidgetSize;
}

void WCFitter::SizeContent(const Vec2<float>& size) const
{
    if (const auto slot = GetSlot(0))
    {
        const auto widget = slot->GetWidget();
        const auto widgetDesiredSize = widget->GetDesiredSize();
        auto newDrawSize = widgetDesiredSize;
        if (!widgetDesiredSize.NearlyEquals(size, 0.001))
        {
            switch (GetMode())
            {
            case FitMode::None:
                break;
            case FitMode::Fill:
                {
                    newDrawSize = size;
                }
                break;
            case FitMode::Contain:
                {
                    newDrawSize = ComputeContainSize(size, widgetDesiredSize);
                }
                break;
            case FitMode::Cover:
                {
                    newDrawSize = ComputeCoverSize(size, widgetDesiredSize);
                }
                break;
            }
        }

        if (!widget->GetSize().NearlyEquals(newDrawSize, 0.001)) widget->SetSize(newDrawSize);

        const auto halfSelfDrawSize = size / 2.0f;
        const auto halfSlotDrawSize = newDrawSize / 2.0f;

        if (const auto diff = halfSelfDrawSize - halfSlotDrawSize; !widget->GetOffset().NearlyEquals(diff, 0.001f))
        {
            widget->SetOffset(diff);
        }
    }
}

File: aerox.Runtime.Widgets/Containers/WCFitter.cs
using aerox.Runtime.Math;

namespace aerox.Runtime.Widgets.Containers;

public enum FitMode
{
    Fill,
    Contain,
    Cover,
    None
}

public class WCFitter : Container
{
    private FitMode _fitFittingMode = FitMode.Fill;
    
    public WCFitter(Widget? widget = null) : base(widget == null ? [] : [widget])
    {
        Clip = ClipMode.Bounds;
    }

    public FitMode FittingMode
    {
        get => _fitFittingMode;
        set
        {
            var old = _fitFittingMode;
            _fitFittingMode = value;
            if (_fitFittingMode != old)
            {
                TryUpdateDesiredSize();
                SizeContent(GetContentSize());
            }
        }
    }

    protected override Size2d ComputeDesiredContentSize()
    {
        if (GetSlot(0) is { } slot)
        {
            return slot.GetWidget().GetDesiredSize();
        }
        return new Size2d();
    }

    public override int GetMaxSlots() => 1;

    public static Size2d ComputeContainSize(Size2d drawSize, Size2d widgetSize)
    {
        var widgetAspect = widgetSize.Height / widgetSize.Width;
        var scaledWidgetSize = new Size2d(drawSize.Width, drawSize.Width * widgetAspect);

        if (drawSize.Equals(scaledWidgetSize)) return scaledWidgetSize;
        
        return scaledWidgetSize.Height <= drawSize.Height
            ? scaledWidgetSize
            : new Size2d(drawSize.Height / widgetAspect, drawSize.Height);
    }

    public static Size2d ComputeCoverSize(Size2d drawSize, Size2d widgetSize)
    {
        var widgetAspect = widgetSize.Height / widgetSize.Width;
        var scaledWidgetSize = new Size2d(drawSize.Width, drawSize.Width * widgetAspect);

        if (drawSize.Equals(scaledWidgetSize)) return scaledWidgetSize;


        return scaledWidgetSize.Height <= drawSize.Height
            ? new Size2d(drawSize.Height / widgetAspect, drawSize.Height)
            : scaledWidgetSize;
    }

    public void SizeContent(Size2d drawSize)
    {
        if (GetSlot(0) is { } slot)
        {
            var widget = slot.GetWidget();
            var widgetSize = widget.GetDesiredSize();
            var newDrawSize = widgetSize.Equals(drawSize)
                ? widgetSize
                : _fitFittingMode switch
                {
                    FitMode.Fill => drawSize,
                    FitMode.Contain => ComputeContainSize(drawSize, widgetSize),
                    FitMode.Cover => ComputeCoverSize(drawSize, widgetSize),
                    FitMode.None => widgetSize,
                    _ => throw new ArgumentOutOfRangeException()
                };

            if (!newDrawSize.Equals(widget.GetContentSize())) widget.SetSize(newDrawSize);


            Vector2<float> halfSelfDrawSize = drawSize;
            halfSelfDrawSize /= 2.0f;
            Vector2<float> halfSlotDrawSize = newDrawSize;
            halfSlotDrawSize /= 2.0f;

            var diff = halfSelfDrawSize - halfSlotDrawSize;

            if (!widget.GetOffset().Equals(diff)) widget.SetOffset(diff);
        }
    }

    protected override void ArrangeSlots(Size2d drawSize)
    {
        SizeContent(drawSize);
    }
    
}